src/Mock.ts

```diff
@@ -30,22 +30,35 @@ export interface CalledWithMock<T, Y extends any[]> extends jest.Mock<T, Y> {
     calledWith: (...args: Y | MatchersOrLiterals<Y>) => jest.Mock<T, Y>;
 }
 
-export type MockProxy<T> = {
-    [K in keyof T]: T[K] extends (...args: infer A) => infer B ? CalledWithMock<B, A> : T[K];
-} &
-    T;
+export type _MockProxy<T> = {
+    [K in keyof T]: T[K] extends (...args: infer A) => infer B
+                    ? T[K] & CalledWithMock<B, A>
+                    : T[K];
+};
+
+export type MockProxy<T> = _MockProxy<T> & T;
 
-export type DeepMockProxy<T> = {
+export type _DeepMockProxy<T> = {
     // This supports deep mocks in the else branch
-    [K in keyof T]: T[K] extends (...args: infer A) => infer B ? CalledWithMock<B, A> : DeepMockProxy<T[K]>;
-} &
```

Comment:
This causes TypeScript to expand and immediately compute the type, potentially doing dangerous eager loading on deep and recursive or mutually self-referencing objects - potentially causing perf issues.